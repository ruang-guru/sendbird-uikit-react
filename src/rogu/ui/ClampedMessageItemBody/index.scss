@import '../../../styles/variables';

.rogu-clamped-message-item-body {
  position: relative;
  display: inline-block;
  box-sizing: content-box;

  .rogu-clamped-message-item-body__message {
    white-space: pre-wrap;
    word-break: break-all;
  }

  .rogu-clamped-message-item-body__read-more > * {
    font-weight: bolder;
  }

  .rogu-clamped-message-item-body__inner {
    overflow: hidden;
    text-overflow: ellipsis;
    display: -webkit-box;
    word-break: break-word;
    white-space: pre-wrap; // preserve white-space and line break
    -webkit-line-clamp: 10;
    -webkit-box-orient: vertical;
  }

  &.rogu-clamped-message-item-body--incoming {
    .rogu-clamped-message-item-body__read-more {
      @include themed() {
        color: t(secondary-3);
      }
    }
  }

  &.rogu-clamped-message-item-body--expanded {
    .rogu-clamped-message-item-body__inner {
      display: block;
      overflow: visible;
    }
  }

  &.rogu-clamped-message-item-body--viewer-mode {
    position: absolute;
    bottom: 5%;
    max-height: 40%;
    min-width: 85%;
    max-width: 95%;
    background: #f1f4f9;
    backdrop-filter: blur(16px);
    padding: 24px;
    overflow: auto;
    transition: all 0.2s;
    display: flex;
    flex-direction: column;

    @include sm {
      max-width: 70%;
    }

    &.rogu-clamped-message-item-body--viewer-mode__hidden {
      max-height: 0;
      padding: 0;
    }
  }

  &.rogu-clamped-message-item-body--preview-mode {
    display: block;
  }
}
