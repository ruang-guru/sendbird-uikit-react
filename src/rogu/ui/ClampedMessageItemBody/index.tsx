import React, {
  useContext,
  useEffect,
  useRef,
  useState,
  ReactElement,
} from 'react';

import TextButton from '../TextButton';
import Label, { LabelTypography, LabelColors } from '../Label';
import { getClassName } from '../../../utils';
import { LocalizationContext } from '../../../lib/LocalizationContext';

import './index.scss';

type ClampType = 'init' | 'clamped' | 'expanded';
type ModeType = 'normal' | 'fileViewerCaption' | 'thumbnailCaption';

interface Props {
  className?: string | Array<string>;
  isByMe?: boolean;
  content: string;
  mode?: ModeType;
  isHidden?: boolean;
}

export default function TextMessageItemBody({
  className,
  isByMe = false,
  content,
  mode = 'normal',
  isHidden = false,
}: Props): ReactElement {
  const { stringSet } = useContext(LocalizationContext);
  const [clampState, setClampState] = useState<ClampType>('init');
  const textRef = useRef<HTMLDivElement>(null);

  useEffect(() => {
    if (
      textRef.current &&
      textRef.current.scrollHeight > textRef.current.clientHeight
    ) {
      setClampState('clamped');
    }
  }, [textRef.current]);

  function handleExpand() {
    setClampState('expanded');
  }

  return (
    <div
      className={getClassName([
        className,
        'rogu-clamped-message-item-body',
        clampState == 'expanded'
          ? 'rogu-clamped-message-item-body--expanded'
          : '',
        !isByMe ? 'rogu-clamped-message-item-body--incoming' : '',
        mode === 'fileViewerCaption'
          ? 'rogu-clamped-message-item-body--viewer-mode'
          : '',
        mode === 'fileViewerCaption' && isHidden
          ? 'rogu-clamped-message-item-body--viewer-mode__hidden'
          : '',
        mode === 'thumbnailCaption'
          ? 'rogu-clamped-message-item-body--preview-mode'
          : '',
      ])}
    >
      <div ref={textRef} className="rogu-clamped-message-item-body__inner">
        {content?.split(/\r/).map((word, i) =>
          word === '' ? (
            <br key={i} />
          ) : (
            <Label
              className="rogu-clamped-message-item-body__message"
              color={LabelColors.ONBACKGROUND_1}
              key={i}
              type={LabelTypography.BODY_1}
            >
              {word}
            </Label>
          )
        )}
      </div>

      {clampState === 'clamped' && (
        <TextButton
          className="rogu-clamped-message-item-body__read-more"
          onClick={handleExpand}
        >
          <Label type={LabelTypography.BODY_1}>
            {stringSet.BUTTON__READ_MORE}
          </Label>
        </TextButton>
      )}
    </div>
  );
}
