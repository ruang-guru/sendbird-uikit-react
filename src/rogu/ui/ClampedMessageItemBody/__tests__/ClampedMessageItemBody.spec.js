import React from 'react';
import renderer from 'react-test-renderer';

import ClampedMessageItemBody from "../index";

describe('ClampedMessageItemBody', () => {
  it('should do a snapshot test of the ClampedMessageItemBody DOM', function() {
    const text = "example-text";
    const component = renderer.create(
      <ClampedMessageItemBody content="Some message here" />,
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
