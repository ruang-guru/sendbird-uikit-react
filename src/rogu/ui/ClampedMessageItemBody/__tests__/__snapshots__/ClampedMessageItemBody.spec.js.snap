// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClampedMessageItemBody should do a snapshot test of the ClampedMessageItemBody DOM 1`] = `
<div
  className=" rogu-clamped-message-item-body  rogu-clamped-message-item-body--incoming   "
>
  <div
    className="rogu-clamped-message-item-body__inner"
  >
    <span
      className="rogu-text-message-item-body__message sendbird-label sendbird-label--body-1 sendbird-label--color-onbackground-1"
    >
      Some message here
    </span>
  </div>
</div>
`;
